### PR TITLE
cadl, support multiple content types - add test case first

### PR DESCRIPTION
**Pull request by haolingdong-msft**
Link https://github.com/Azure/autorest.java/issues/1958

First add two cases, to verify existing behavior.
1. data type is not union, contentType is union.
2. data type is union, contentType is union, with overloads to define the mapping.

Other cases worthwhile to add:
- multi content type with LRO and paging
- multi content type with @convenientApi and @protocolApi
- shared route

### CODE REVIEW

**Comment by weidongxu-microsoft** [comment]
The protocol method seems good. But please still do a verification of make REST call with them and inspect the payload.

The overload part is a bit unexpected, need more info on how emitter handles it and how we had/should handle afterwards.

We might need to check with Srikanta whether we should map all Cadl bytes to `BinaryData` (instead of `byte[]`).

You can actually set convenientAPI=false in Cadl source and check it in, if the protocol method is good.

**Comment by weidongxu-microsoft** [comment]
though you probably need to handle typespec upgrade/rename first...

**Comment by haolingdong-msft** [comment]
For the issues you mentioned above, I have below thoughts, please let me know your comments/suggestions.

1. map `byte[]` to `BinaryData`:
I think we can map for multiple content types case to `BinaryData` only, e.g. emitter can set the schema type as `binary` in code-model.yaml like below. We can then add logic to set return type as `BinaryData` for [`Schema.AllSchemaTypes.BINARY`](https://github.com/Azure/autorest.java/blob/main/extension-base/src/main/java/com/azure/autorest/extension/base/model/codemodel/Schema.java#L276) type in [CadlPrimitiveMapper.java](https://github.com/Azure/autorest.java/blob/main/cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java#L22)
image

2. expect upload in proxy method, but we got 3:
I think we can add a flag in `code-model.yaml` called `overload: boolean` to indicate the operation is overload operation or not. If it is overload operation, in codegen, we don't need to generate protocol method and proxy method for it.
```yaml
...
generateProtocolApi: true
overload: true
```

**Comment by weidongxu-microsoft** [comment]
1.
That might do, does m4 do this to request schema?

2.
Probably further discussion is needed. `overload` won't tell you which API to override.
Alternative could be customize the convenicenceAPI.requests (it can put multiple requests as different overload to convenience API)

**Comment by haolingdong-msft** [comment]
1. M4 has `requestMediaTypes`. This is an example m4 output:
image
image


2. I understand your alternative. The structure of code-model would better describe the relationship between major operation and overload operations. I will think more of this alternative.
What I try to propose previously is: If we have 1 major operation with 3 overload operations. The compiler will simply return 4 operations regardless it's overload or not. That's why we generate both protocol methods and convenient methods for the 4 operations. I'm thinking if we could add `overload: true` to the 3 overload operations in `code-model.yaml`. So the codegen can disable generation for proxy methods and protocol methods for those operations. We will also put name (in this case `upload`) of the operation into `convenicenceAPI`, so the convenient method name will be changed.

**Comment by weidongxu-microsoft** [comment]
The problem is when you disable the proxy method and protocol method, the previous logic of convenience method calling corresponding protocol method will likely won't work (of course, even if we reuse it, it might still need some enhancement). So this might result in you writing a new logic of convenience method calling whatever, for overload.

**Comment by weidongxu-microsoft** [comment]
For 1, you can first check/fix in in a new PR, if this happens to single content-type=octect-stream as well.

File: cadl-extension/src/code-model-builder.ts [hunk 1]
```diff
@@ -701,7 +701,16 @@ export class CodeModelBuilder {
...
+    let schema = this.processSchema(body.kind === "Model" ? body : body.type, body.name);
+    // handle for binary body
+    if (
+      parameters.properties.has("contentType") &&
+      body.kind === "ModelProperty" &&
+      body.type.kind === "Scalar" &&
+      body.type.name === "bytes"
+    ) {
+      schema = new BinarySchema(body.name);
+    }
```
Comment (weidongxu-microsoft) [inline cadl-extension/src/code-model-builder.ts:713]: 1. should be an `if else` for the schema
2. do we need to test "contentType", probably just drop it (and even if we need to test, "contentType" is not a wire name; also then we would need to test again "application/json" or anything relates to JSON, as it is totally valid for user to write `content-type: application/json` even it is not necessary)?

Comment (weidongxu-microsoft) [reply cadl-extension/src/code-model-builder.ts:713]: If user uses `application/json` and implicit body (as we most use here), the body would be `kind=Model`

(want to make sure we still can distinguish `@body data: bytes`, from `data: bytes` in an implicit Model for JSON)

File: cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java [hunk 2]
```diff
@@ -21,6 +21,8 @@ public static PrimitiveMapper getInstance() {
     protected IType createPrimitiveType(PrimitiveSchema primaryType) {
         if (primaryType.getType() == Schema.AllSchemaTypes.DATE) {
             return ClassType.LocalDate;
+        } else if (primaryType.getType() == Schema.AllSchemaTypes.BINARY){
+            return ClassType.BinaryData;
```
Comment (weidongxu-microsoft) [inline cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java:25]: `Schema.AllSchemaTypes.BINARY` seems not processed in `PrimitiveMapper`. How does it handle binary date from m4?

Comment (haolingdong-msft) [reply cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java:25]: This is M4 example, it has `requestMediaTypes` on each operation, `requestMediaTypes` will map different content types to different input schema.
image
image
image

It is processed here: https://github.com/Azure/autorest.java/blob/main/javagen/src/main/java/com/azure/autorest/mapper/MethodGroupMapper.java#L89

Comment (weidongxu-microsoft) [reply cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java:25]: How it process that, I cannot find code `requestMediaTypes` in the whole project?

Comment (weidongxu-microsoft) [reply cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java:25]: E.g. in vanilla mediatypes test, it has
```java
    @ServiceMethod(returns = ReturnType.SINGLE)
    public String analyzeBody(ContentType contentType, Flux<ByteBuffer> input, Long contentLength) {
        return analyzeBodyWithResponse(contentType, input, contentLength, Context.NONE).getValue();
    }
```

Ignore the `contentLength` (DPG does not have it), it has the `Flux<ByteBuffer> input`. I assumed something mapped to `Flux<ByteBuffer>`?

Comment (weidongxu-microsoft) [reply cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java:25]: Also, we are talking about single content-type with binary upload.

We can avoid discussion of multiple content-type for now.

Comment (weidongxu-microsoft) [reply cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java:25]: > image requests refer to the item in requestMediaTypes, we proccess requests.

I see. But still this question. How vanilla process this without mapping in `PrimitiveMapper`? Your code is DPG only (which map any payload to BinaryData)

Comment (haolingdong-msft) [reply cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java:25]: This is where we map `BinarySchema` to `FluxByteBuffer`
https://github.com/Azure/autorest.java/blob/main/javagen/src/main/java/com/azure/autorest/mapper/BinaryMapper.java#L28

call stack:
image

Comment (weidongxu-microsoft) [reply cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java:25]: So this change here is not needed?

Comment (haolingdong-msft) [reply cadl-extension/src/main/java/com/azure/cadl/mapper/CadlPrimitiveMapper.java:25]: I think so..I'm testing..

File: cadl-extension/src/code-model-builder.ts (hunk 1 above)
Comment (weidongxu-microsoft) [reply cadl-extension/src/code-model-builder.ts]: Use `if else`. Don't process schema you do not use.

File: cadl-extension/src/code-model-builder.ts [hunk 3]
```diff
@@ -701,7 +701,13 @@ export class CodeModelBuilder {
   }
 
   private processParameterBody(op: CodeModelOperation, body: ModelProperty | Model, parameters: Model) {
-    const schema = this.processSchema(body.kind === "Model" ? body : body.type, body.name);
+    let schema: Schema;
+    if (body.kind === "ModelProperty" && body.type.kind === "Scalar" && body.type.name === "bytes") {
+      //handle for binary body
+      schema = new BinarySchema(body.name);
```
Comment (weidongxu-microsoft) [inline cadl-extension/src/code-model-builder.ts:707]: Ideally this should be a `processBinary` which will add it to `codeModel.schemas` (that's the side-effect)
https://github.com/Azure/autorest.java/blob/54b1738de6d48b23a298fda4581198c2b3af2815/cadl-extension/src/code-model-builder.ts#L1142-L1148

Recommend to do this.

**Review by weidongxu-microsoft — APPROVED** [review]

**Pull request merged**